src/storage/FileDescriptor.cpp

```diff
@@ -178,76 +180,180 @@ FileDescriptor FileDescriptor::stderrFactory()
     }
 
-    // Call readdir to get next dirent
-    struct dirent* direntPtr = ::readdir(dirPtr);
+    // Load all directory entries
+    uint64_t nextIdx = 0;
+    struct dirent* direntPtr;
+    while ((direntPtr = ::readdir(dirPtr)) != nullptr) {
+        nextIdx++;
 
-    // Build the actual dirent
-    DirEnt d;
-    if (!direntPtr) {
-        // Close iterator
-        closedir(dirPtr);
-        iterFinished = true;
-    } else {
-        d.type = direntPtr->d_type;
-        d.ino = direntPtr->d_ino;
-        d.path = std::string(direntPtr->d_name);
+        DirEnt nextEnt;
 
         // We have to set "next" here to specify the offset of this
         // directory entry. It seems this is only used to be passed
         // back as the "cookie" value to fd_readdir
-        d.next = direntPtr->d_off;
+        nextEnt.next = nextIdx;
+
+        nextEnt.type = direntPtr->d_type;
+        nextEnt.ino = direntPtr->d_ino;
+        nextEnt.path = std::string(direntPtr->d_name);
+
+        dirContents.push_back(nextEnt);
+    }
+
+    // Close iterator
+    closedir(dirPtr);
+    logger->debug("Loaded {} entries for {}", dirContents.size(), realPath);
+
+    // Set flag
+    dirContentsLoaded = true;
+}
+
+void FileDescriptor::iterBack()
+{
+    if (!dirContentsLoaded) {
+        throw std::runtime_error("Iterator not started, cannot go back");
     }
 
-    return d;
+    if (dirContentsIdx == 0) {
+        throw std::runtime_error("Iterator already at zero, cannot go back");
+    }
+
+    dirContentsIdx--;
+}
+
+bool FileDescriptor::iterStarted()
+{
+    return dirContentsLoaded;
+}
+
+bool FileDescriptor::iterFinished()
+{
+    return dirContentsLoaded && (dirContentsIdx >= dirContents.size());
+}
+
+DirEnt FileDescriptor::iterNext()
+{
+    // Check if contents are initialised
+    if (!dirContentsLoaded) {
+        loadDirContents();
+    }
+
+    // Check
+    if (iterFinished()) {
+        throw std::runtime_error(
+          fmt::format("Accessing index {} in directory length {}",
+                      dirContentsIdx,
+                      dirContents.size()));
+    }
+
+    DirEnt nextEntry = dirContents.at(dirContentsIdx);
+
+    // Increment the iterator
+    dirContentsIdx++;
+
+    return nextEntry;
+}
+
+/**
+ * Note that this function conforms to the standard readdir interface:
+ * - Copy each dirent struct followed by its path string
+ * - If the last path string doesn't fit, return the buffer length
+ * - If the last dirent doesn't fit, don't copy it and return the buffer length
+ *
+ * The external libc wrapper should prevent this third point.
+ *
+ * If the final dirent or path doesn't fit, the caller will expect the next
+ * result to contain that same dirent/ path, so we have to step back one in the
+ * iterator if that's the case.
+ *
+ * The caller knows they've reached the end when the buffer is _not_ filled by
+ * this call.
+ */
+size_t FileDescriptor::copyDirentsToWasiBuffer(uint8_t* buffer,
```

Comment:
The logic here is the stuff we need to get right for everything else to work. It used to be held in `src/wavm/io.cpp` but it's now captured here and tested as part of this class.

Reply:
Yep, this is what the libc wrapper expects